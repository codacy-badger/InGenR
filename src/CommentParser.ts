import { isNumber } from "./validators"
import { Readable } from "stream"
import * as EventEmitter from "events"
import * as fs from "fs"
import * as rl from "readline"
import * as yaml from "js-yaml"
import _debug from "debug"
import { WarningEntry, warnInterpolated, warnInvalidArgBody } from "./warnings"

const debug = _debug("InGenR:CommentParser")

export const defaultParserOptions = {
  commentStartRegex: "\\/\\*!",
  commentEndRegex: "\\*\\/",
  commentLBoundRegEx: "^\\s*\\*?",
  commentRBoundRegEx: "\\*?\\s*$"
}

interface Matchers {
  blockStartRegex: RegExp
  blockEndRegex: RegExp
  blockArgsBodyLineRegex: RegExp
  commentEndRegex: RegExp
}

export interface Directive {
  templateName: string
  templateArgs: any
  directiveArgs: {
    targetFilePath: string
  } | undefined
  blockStartLineIndex: number
  blockEndLineIndex: number
  currentContent?: string[]
  compiledContent?: string
}

export interface CandidateBlock extends Partial<Directive> {
  argsLines: string[]
  contentLines: string[]
}

enum ParserState {
  INIT = "INIT",
  IN_ARGS_BLOCK = "IN_ARGS_BLOCK",
  IN_GENERATED_BLOCK = "IN_GENERATED_BLOCK"
}

export class CommentParser extends EventEmitter {
  private matchers: Matchers
  private currentCandidate: CandidateBlock | null = null
  constructor(public inputStream: Readable, private parseOptions = defaultParserOptions) {
    super()
    this.matchers = {
      blockStartRegex: new RegExp(
        `${parseOptions.commentStartRegex}\\s+InGenR:expand\\s+(\\S*)\\s*($|${
          parseOptions.commentEndRegex
        })`
      ),
      blockEndRegex: new RegExp(
        `${parseOptions.commentStartRegex}\\s+InGenR:end\\s*${parseOptions.commentEndRegex}`
      ),
      blockArgsBodyLineRegex: new RegExp(
        `${parseOptions.commentLBoundRegEx}(.*)${parseOptions.commentRBoundRegEx}`
      ),
      commentEndRegex: new RegExp(`${parseOptions.commentEndRegex}`)
    }
    debug("Matchers:", this.matchers)
  }

  get parserState() {
    if (!this.currentCandidate) {
      return ParserState.INIT
    } else if (isNumber(this.currentCandidate.blockStartLineIndex)) {
      if (isNumber(this.currentCandidate.blockEndLineIndex)) {
        throw new Error("Invalid state encountered")
      }
      return ParserState.IN_GENERATED_BLOCK
    } else {
      return ParserState.IN_ARGS_BLOCK
    }
  }

  private parseLine(lineIndex: number, line: string) {
    debug("[ParserState: %s] Processing line %d: %s", this.parserState, lineIndex, line)
    const prevParserState = this.parserState
    let warnings: WarningEntry[] = []
    switch (prevParserState) {
      case ParserState.INIT:
        this.checkBlockStart(lineIndex, line, warnings)
        this.emitLine(lineIndex, line, prevParserState, warnings)
        break
      case ParserState.IN_ARGS_BLOCK:
        this.accumulateArgs(lineIndex, line, warnings)
        this.emitLine(lineIndex, line, prevParserState, warnings)
        break
      case ParserState.IN_GENERATED_BLOCK:
        this.checkGeneratedEnd(lineIndex, line, warnings)
        // @ts-ignore
        if (this.parserState === ParserState.INIT) {
          this.emitLine(lineIndex, line, prevParserState, warnings)
        }
    }
  }

  private emitLine(
    lineIndex: number,
    line: string,
    prevParserState: ParserState,
    warnings: WarningEntry[]
  ) {
    const unprocessed = prevParserState === this.parserState
    if (unprocessed) {
      this.checkUnprocessedDirectives(line, warnings)
    }
    this.emit("item", {
      type: "LINE",
      lineIndex,
      data: line,
      parserState: this.parserState,
      didProcess: !unprocessed,
      warnings
    })
  }

  private checkUnprocessedDirectives(line: string, warnings: WarningEntry[]) {
    const match = line.match(/InGenR:\S+/)
    if (!match) return
    warnings.push({
      index: match.index,
      message:
        `Potentially unprocessed InGenR directive: ${match[0]}.\n` +
        `If you expected this directive to be processed please recheck the syntax`
    })
  }

  private checkBlockStart(lineIndex: number, line: string, warnings: WarningEntry[]) {
    const match = line.match(this.matchers.blockStartRegex)
    if (!match || !match[1]) return
    if (match[0].trim().length !== line.trim().length) {
      warnings.push(warnInterpolated(match.index))
      return
    }
    this.currentCandidate = {
      templateName: match[1],
      argsLines: [],
      contentLines: []
    }
    if (match[2]) {
      // Closes on the same line
      this.currentCandidate.blockStartLineIndex = lineIndex + 1
    }
  }

  private accumulateArgs(lineIndex: number, line: string, warnings: WarningEntry[]) {
    const match = line.match(this.matchers.blockArgsBodyLineRegex)
    if (line.match(this.matchers.commentEndRegex)) {
      this.currentCandidate!.blockStartLineIndex = lineIndex + 1
      return
    }
    if (match) {
      this.currentCandidate!.argsLines.push(match[1])
      return
    }
    warnings.push(warnInvalidArgBody())
  }

  private checkGeneratedEnd(lineIndex: number, line: string, warnings: WarningEntry[]) {
    const match = line.match(this.matchers.blockEndRegex)
    if (!match) {
      this.currentCandidate!.contentLines.push(line)
      return
    }
    if (match[0].trim().length !== line.trim().length) {
      warnings.push(warnInterpolated(match.index))
      return
    }
    let templateArgs
    let directiveArgs
    try {
      ({templateArgs, directiveArgs} = this.parseArgs())
    } catch (e) {
      warnings.push({
        message: "Failed to parse template arguments. This directive will be discarded."
      })
      return
    }
    this.emit("item", {
      type: "PARSED_BLOCK",
      data: {
        templateName: this.currentCandidate!.templateName!,
        templateArgs,
        directiveArgs,
        blockStartLineIndex: this.currentCandidate!.blockStartLineIndex!,
        blockEndLineIndex: lineIndex,
        currentContent: this.currentCandidate!.contentLines
      },
      parserState: this.parserState
    })
    this.currentCandidate = null
  }

  private parseArgs() {
    const {argsLines} = this.currentCandidate!
    const separatorIndex = argsLines.findIndex(line => line.trim() === "---")
    let templateArgsBody: string | undefined
    let directiveArgsBody: string | undefined
    if (separatorIndex >= 0) {
      directiveArgsBody = argsLines.slice(0, separatorIndex).join("\n")
      templateArgsBody = argsLines.slice(separatorIndex + 1).join("\n")
    } else {
      templateArgsBody = argsLines.join("\n")
    }
    const result = {
      templateArgs: this.parseArgsBody(templateArgsBody),
      directiveArgs: this.parseArgsBody(directiveArgsBody)
    }
    debug("parsed directive args:", result)
    return result
  }

  private parseArgsBody(body: string | undefined) {
    if (!body) return null
    debug("Loading as yaml: %s", body)
    return yaml.safeLoad(body)
  }

  parse(): Promise<void> {
    return new Promise((resolve, reject) => {
      const lineReader = rl.createInterface({
        input: this.inputStream
      })
      let lineIndex = 0
      lineReader.on("line", line => {
        this.parseLine(lineIndex, line)
        lineIndex++
      })
      lineReader.on("error", e => reject(e))
      lineReader.on("close", () => {
        debug("[ParserState: %s] close", this.parserState)
        this.emit("close")
        if (this.parserState !== ParserState.INIT) {
          reject(new Error(`Unexpected End of File while parsing`))
          return
        }
        resolve()
      })
    })
  }
}
