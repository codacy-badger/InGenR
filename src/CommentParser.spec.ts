// @ts-ignore
import * as BufferStream from "node-bufferstream"
import { CommentParser } from "./CommentParser"

const parseContent = async (content: string) => {
  const stream = new BufferStream()
  const parser = new CommentParser(stream)
  const items: any[] = []
  parser.on("item", (item: any) => {
    items.push(item)
  })
  const promise = parser.parse()
  stream.write(content)
  await promise
  return items
}

describe("Comment Parser", () => {
  it("parses empty file", async () => {
    const items = await parseContent("")
    expect(items.length).toEqual(0)
  })

  it("parses files with no comment blocks", async () => {
    const items = await parseContent(`
            import express from "express";
            const app = express();
            app.listen()
        `)
    expect(items).toMatchSnapshot()
  })

  it("parses files with comment blocks", async () => {
    const items = await parseContent(`
            /**! InGenR:expand sample */
            /**! InGenR:end */

            /**! InGenR:expand knex-dal
            *
            * tableName: users
            * columns:
            *   - name: name
            *     type: string
            *   - name: email
            *     type: string
            */
            /**! InGenR:end */

            /**! InGenR:expand products
             *
             * tableName: products
             */
            /**! InGenR:end */
        `)
    expect(items).toMatchSnapshot()
  })
})
