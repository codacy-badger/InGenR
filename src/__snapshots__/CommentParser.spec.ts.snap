// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Parser parses files with comment blocks 1`] = `
Array [
  Object {
    "data": "",
    "didProcess": false,
    "lineIndex": 0,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            /**! InGenR:expand sample */",
    "didProcess": false,
    "lineIndex": 1,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:expand.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "            /**! InGenR:end */",
    "didProcess": false,
    "lineIndex": 2,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:end.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "",
    "didProcess": false,
    "lineIndex": 3,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            /**! InGenR:expand knex-dal",
    "didProcess": false,
    "lineIndex": 4,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:expand.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "            *",
    "didProcess": false,
    "lineIndex": 5,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            * tableName: users",
    "didProcess": false,
    "lineIndex": 6,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            * columns:",
    "didProcess": false,
    "lineIndex": 7,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            *   - name: name",
    "didProcess": false,
    "lineIndex": 8,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            *     type: string",
    "didProcess": false,
    "lineIndex": 9,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            *   - name: email",
    "didProcess": false,
    "lineIndex": 10,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            *     type: string",
    "didProcess": false,
    "lineIndex": 11,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            */",
    "didProcess": false,
    "lineIndex": 12,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            /**! InGenR:end */",
    "didProcess": false,
    "lineIndex": 13,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:end.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "",
    "didProcess": false,
    "lineIndex": 14,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            /**! InGenR:expand products",
    "didProcess": false,
    "lineIndex": 15,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:expand.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "             *",
    "didProcess": false,
    "lineIndex": 16,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "             * tableName: products",
    "didProcess": false,
    "lineIndex": 17,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "             */",
    "didProcess": false,
    "lineIndex": 18,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            /**! InGenR:end */",
    "didProcess": false,
    "lineIndex": 19,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [
      Object {
        "index": 17,
        "message": "Potentially unprocessed InGenR directive: InGenR:end.
If you expected this directive to be processed please recheck the syntax",
      },
    ],
  },
  Object {
    "data": "        ",
    "didProcess": false,
    "lineIndex": 20,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
]
`;

exports[`Comment Parser parses files with no comment blocks 1`] = `
Array [
  Object {
    "data": "",
    "didProcess": false,
    "lineIndex": 0,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            import express from \\"express\\";",
    "didProcess": false,
    "lineIndex": 1,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            const app = express();",
    "didProcess": false,
    "lineIndex": 2,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "            app.listen()",
    "didProcess": false,
    "lineIndex": 3,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
  Object {
    "data": "        ",
    "didProcess": false,
    "lineIndex": 4,
    "parserState": "INIT",
    "type": "LINE",
    "warnings": Array [],
  },
]
`;
